bundles/org.openhab.binding.playstation/src/main/java/org/openhab/binding/playstation/internal/PS4Handler.java

```diff
@@ -0,0 +1,821 @@
+        } catch (InterruptedException e) {
+            // Ignore, sleep just to make keyEvents behave better.
+        }
+        return true;
+    }
+
+    private boolean login() {
+        try {
+            SocketChannel channel = getConnection(false);
+            return login(channel);
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            logger.info("Send packet exception: {}", e.getMessage());
+        }
+        return false;
+    }
+
+    private void logOut() {
+        ByteBuffer outPacket = PS4PacketHandler.makeLogoutPacket();
+        sendPacketEncrypted(outPacket);
+    }
+
+    private void takeScreenShot() {
+        ByteBuffer outPacket = PS4PacketHandler.makeScreenShotPacket();
+        sendPacketEncrypted(outPacket);
+    }
+
+    /**
+     * This closes the connection with the PS4.
+     */
+    private void sendByeBye() {
+        ByteBuffer outPacket = PS4PacketHandler.makeByebyePacket();
+        sendPacketEncrypted(outPacket, false);
+    }
+
+    private void turnOnPS4() {
+        wakeUpPS4();
+        scheduler.schedule(this::waitAndConnectToPS4, 13, TimeUnit.SECONDS);
+    }
+
+    private void waitAndConnectToPS4() {
+        try {
+            getConnection();
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            logger.debug("Turn on PS4 exception: {}", e.getMessage());
+        }
+    }
+
+    private void sendStandby() {
+        ByteBuffer outPacket = PS4PacketHandler.makeStandbyPacket();
+        sendPacketEncrypted(outPacket);
+    }
+
+    /**
+     * Ask PS4 if the OSK is open so we can get and set text.
+     */
+    private void sendOSKStart() {
+        ByteBuffer outPacket = PS4PacketHandler.makeOSKStartPacket();
+        sendPacketEncrypted(outPacket);
+    }
+
+    /**
+     * Sets the entire OSK string on the PS4.
+     *
+     * @param text The text to set in the OSK.
+     */
+    private void setOSKText(String text) {
+        logger.debug("Sending osk text packet,\"{}\"", text);
+        ByteBuffer outPacket = PS4PacketHandler.makeOSKStringChangePacket(text);
+        sendPacketEncrypted(outPacket);
+    }
+
+    /**
+     * Tries to start an application on the PS4.
+     *
+     * @param applicationId The unique id for the application (CUSAxxxxx).
+     */
+    private void startApplication(String applicationId) {
+        ByteBuffer outPacket = PS4PacketHandler.makeApplicationPacket(applicationId);
+        sendPacketEncrypted(outPacket);
+    }
+
+    private void sendRemoteKey(int pushedKey) {
+        try {
+            SocketChannel channel = getConnection();
+
+            ByteBuffer outPacket = PS4PacketHandler.makeRemoteControlPacket(PS4_KEY_OPEN_RC);
+            sendPacketEncrypted(outPacket, channel);
+            Thread.sleep(MIN_SENDKEY_DELAY);
```

Comment:
You should schedule the next step instead of blocking with `Thread.sleep`

```
sendPacketEncrypted(...);
scheduler.schedule(() -> {
   <sendremotekey...>
   scheduler.schedule(() -> {
       <nextStep...>
    }, MIN_SENDKEY_DELAY, TimeUnit...);
}, ...);
```

(just to give you an idea). With JDK9 (i.e.  OH3) `CompletableFuture`s with `delayedExecutor` whpuld be a perfect match IMO. But then we will not get this merged before OH3 release

Reply:
Fixed.